File: app/system/actors/ExternalExamExpirationActor.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package system.actors;

import akka.actor.AbstractActor;
import io.ebean.Ebean;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Collections;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;
import javax.inject.Inject;
import models.Attachment;
import models.json.ExternalExam;
import models.questions.EssayAnswer;
import models.sections.ExamSectionQuestion;
import play.Logger;
import play.libs.ws.WSClient;
import util.config.ConfigReader;

public class ExternalExamExpirationActor extends AbstractActor {

    private static final Logger.ALogger logger = Logger.of(ExternalExamExpirationActor.class);
    private static final int MONTHS_UNTIL_EXPIRATION = 4;

    private final ConfigReader configReader;
    private final WSClient wsClient;

    @Inject
    public ExternalExamExpirationActor(ConfigReader configReader, WSClient wsClient) {
        this.configReader = configReader;
        this.wsClient = wsClient;
    }

    private Optional<URL> parseUrl(String id) {
        String url = configReader.getIopHost() + "/api/attachments/" + id;
        try {
            return Optional.of(new URL(url));
        } catch (MalformedURLException e) {
            return Optional.empty();
        }
    }

    private CompletableFuture<Void> deleteAttachment(Attachment attachment) {
        Optional<URL> url = parseUrl(attachment.getExternalId());
        return CompletableFuture.runAsync(() -> url.ifPresent(value -> wsClient.url(value.toString()).delete()));
    }

    @Override
    public Receive createReceive() {
        return receiveBuilder()
            .match(
                String.class,
                s -> {
                    logger.debug("Starting external exam expiration check ->");
                    Set<ExternalExam> exams = Ebean
                        .find(ExternalExam.class)
                        .where()
                        .isNotNull("sent")
                        .jsonExists("content", "id")
                        .findSet();

                    for (ExternalExam ee : exams) {
                        if (ee.getSent() == null) {
                            continue;
                        }
                        Set<Attachment> attachments = ee
                            .deserialize()
                            .getExamSections()
                            .stream()
                            .flatMap(es -> es.getSectionQuestions().stream())
                            .map(ExamSectionQuestion::getEssayAnswer)
                            .filter(ea -> ea != null && ea.getAttachment() != null)
                            .map(EssayAnswer::getAttachment)
                            .collect(Collectors.toSet());
                        CompletableFuture
                            .allOf(attachments.stream().map(this::deleteAttachment).toArray(CompletableFuture[]::new))
                            .thenRunAsync(() -> {
                                if (ee.getSent().plusMonths(MONTHS_UNTIL_EXPIRATION).isBeforeNow()) {
                                    ee.setContent(Collections.emptyMap());
                                    ee.update();
                                    logger.info("Marked external exam {} as expired", ee.getId());
                                }
                            });
                    }
                    logger.debug("<- done");
                }
            )
            .build();
    }
}

File: app/system/actors/ExternalExamExpirationActor.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package system.actors

import io.ebean.DB
import miscellaneous.config.ConfigReader
import miscellaneous.scala.DbApiHelper
import models.attachment.Attachment
import models.iop.ExternalExam
import org.apache.pekko.actor.AbstractActor
import org.apache.pekko.stream.Materializer
import org.apache.pekko.stream.scaladsl.{Sink, Source}
import play.api.Logging
import play.api.libs.ws.WSClient

import java.net.{MalformedURLException, URI}
import javax.inject.Inject
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters.*
import scala.util.control.Exception.catching
import scala.util.{Failure, Success}

object ExternalExamExpirationActor:
  private val MONTHS_UNTIL_EXPIRATION = 4
class ExternalExamExpirationActor @Inject (
    private val configReader: ConfigReader,
    private val wsClient: WSClient,
    implicit val ec: ExecutionContext,
    implicit val mat: Materializer
) extends AbstractActor
    with DbApiHelper
    with Logging:

  // Maximum number of concurrent HTTP requests
  private val maxConcurrency = 10

  private def parseUrl(id: String) =
    val url = s"${configReader.getIopHost}/api/attachments/$id"
    catching(classOf[MalformedURLException]).either(URI.create(url).toURL) match
      case Left(_)  => None
      case Right(u) => Some(u)

  private def deleteAttachment(attachment: Attachment): Future[Unit] =
    val url = parseUrl(attachment.getExternalId).map(_.toString).getOrElse("")
    if url.nonEmpty then
      wsClient
        .url(url)
        .delete()
        .map(_ => ())
        .recover { case e: Exception =>
          logger.error(s"Failed to delete attachment ${attachment.getExternalId}", e)
        }
    else Future.successful(())

  private def processExternalExam(ee: ExternalExam): Future[Unit] =
    val attachments =
      try
        ee.deserialize.getExamSections.asScala
          .flatMap(_.getSectionQuestions.asScala)
          .map(_.getEssayAnswer)
          .filter(ea => Option(ea).nonEmpty && Option(ea.getAttachment).nonEmpty)
          .map(_.getAttachment)
          .toList
      catch
        case e: Exception =>
          logger.error(s"Failed to deserialize external exam ${ee.getId}", e)
          List.empty[Attachment]

    if attachments.nonEmpty then
      val count = attachments.size
      logger.info(
        s"Processing $count attachments for external exam ${ee.getId} with max concurrency of $maxConcurrency"
      )
      Source(attachments)
        .mapAsync(maxConcurrency)(deleteAttachment)
        .runWith(Sink.ignore)
        .map(_ =>
          if ee.getSent.plusMonths(ExternalExamExpirationActor.MONTHS_UNTIL_EXPIRATION).isBeforeNow then
            ee.setContent(Map.empty[String, Object].asJava)
            ee.update()
            logger.info(s"Marked external exam ${ee.getId} as expired")
        )
        .recover { case e: Exception =>
          logger.error(s"Failed in deleting attachments for external exam ${ee.getId}", e)
        }
    else
      if ee.getSent.plusMonths(ExternalExamExpirationActor.MONTHS_UNTIL_EXPIRATION).isBeforeNow then
        ee.setContent(Map.empty[String, Object].asJava)
        ee.update()
        logger.info(s"Marked external exam ${ee.getId} as expired")
      Future.successful(())

  override def createReceive(): AbstractActor.Receive = receiveBuilder()
    .`match`(
      classOf[String],
      (_: String) =>
        logger.info("Starting external exam expiration check ->")
        val externalExams = DB
          .find(classOf[ExternalExam])
          .where
          .isNotNull("sent")
          .jsonExists("content", "id")
          .distinct
          .filter(ee => Option(ee.getSent).nonEmpty)
          .toList

        val count = externalExams.size
        if count > 0 then
          logger.info(s"Processing $count external exams with max concurrency of $maxConcurrency")
          Source(externalExams)
            .mapAsync(maxConcurrency)(processExternalExam)
            .runWith(Sink.ignore)
            .onComplete {
              case Success(_) =>
                logger.info("<- External exam expiration check completed successfully")
              case Failure(e) =>
                logger.error("Error processing external exam expiration", e)
            }
        else
          logger.info("No external exams to process")
          logger.info("<- done")
    )
    .build
